pymc_marketing/mmm/delayed_saturated_mmm.py

```diff
@@ -1471,11 +1823,377 @@ def add_lift_test_measurements(
+                    v[0] / self.channel_transformer["scaler"].scale_.max(),
+                    v[1] / self.channel_transformer["scaler"].scale_.max(),
+                )
+                for k, v in budget_bounds.items()
+            }
+        else:
+            scale_budget_bounds = None
+
+        allocator = BudgetOptimizer(
+            adstock=self.adstock,
+            saturation=self.saturation,
+            parameters=parameters_mid,
+            adstock_first=self.adstock_first,
+            num_days=num_days,
+        )
+
+        self.optimal_allocation_dict, _ = allocator.allocate_budget(
+            total_budget=scale_budget,
+            budget_bounds=scale_budget_bounds,
+            custom_constraints=custom_constraints,
+        )
+
+        inverse_scaled_channel_spend = self.channel_transformer.inverse_transform(
+            np.array([list(self.optimal_allocation_dict.values())])
+        )
+        original_scale_allocation_dict = {
+            k: v
+            for k, v in zip(
+                self.optimal_allocation_dict.keys(),
+                inverse_scaled_channel_spend[0],
+                strict=False,
+            )
+        }
+
+        synth_dataset = self._create_synth_dataset(
+            df=self.X,
+            date_column=self.date_column,
+            allocation_strategy=original_scale_allocation_dict,
+            channels=self.channel_columns,
+            controls=self.control_columns,
+            target_col=self.output_var,
+            time_granularity=time_granularity,
+            time_length=num_days,
+            lag=self.adstock.l_max,
+        )
+
+        synth_dataset[self.date_column] = pd.to_datetime(
+            synth_dataset[self.date_column]
+        )
+
+        return self.sample_posterior_predictive(
+            X_pred=synth_dataset,
+            extend_idata=False,
+            include_last_observations=True,
+            original_scale=False,
+            var_names=["y", "channel_contributions"],
+            progressbar=False,
+        )
+
+    def plot_budget_allocation(
+        self,
+        samples: az.InferenceData,
+        figsize: tuple[float, float] = (12, 6),
+        ax: plt.Axes | None = None,
+        original_scale: bool = True,
+    ):
+        """
+        Plot the budget allocation and channel contributions.
+
+        Parameters
+        ----------
+        samples : az.InferenceData
+            The inference data containing the channel contributions.
+        figsize : tuple[float, float], optional
+            The size of the figure to be created, by default (12, 6).
+        original_scale : bool, optional
+            A boolean flag to determine if the values should be plotted in their original scale,
+            by default True.
+
+        Returns
+        -------
+        tuple
+            A tuple containing the matplotlib figure and axes of the plot.
+        """
+
+        # def plot_budget_allocation(self, samples, ax, original_scale=True):
```

Comment:
remove commented code

Reply:
We often just return the figure ... can we also do it here for consistency?

Reply:
Done majority but the function `plot_budget_allocation` must return both to be able to iterate and plot different created scenarios as in the notebook. 

All the others are following the same style.